From c0ece8d8c6eeea6eea5533bef1a3e76378e211d8 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:18:57 -0400
Subject: [net] netfilter: nf_tables: fix wrong type in transaction when replacing rules

Message-id: <20141021161857.22888.89251.stgit@localhost.localdomain>
Patchwork-id: 97968
O-Subject: [RHEL7 net PATCH 071/124] netfilter: nf_tables: fix wrong type in transaction when replacing rules
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: ac904ac835ac7879a9374dc3ef1e5cb75d9c7ceb

 commit ac904ac835ac7879a9374dc3ef1e5cb75d9c7ceb
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Tue Jun 10 10:53:03 2014 +0200

    netfilter: nf_tables: fix wrong type in transaction when replacing rules

    In b380e5c ("netfilter: nf_tables: add message type to transactions"),
    I used the wrong message type in the rule replacement case. The rule
    that is replaced needs to be handled as a deleted rule.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 19f438d..39369ea 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -1792,7 +1792,7 @@ static int nf_tables_newrule(struct sock *nlsk, struct sk_buff *skb,
 
  if (nlh->nlmsg_flags & NLM_F_REPLACE) {
   if (nft_rule_is_active_next(net, old_rule)) {
-   trans = nft_trans_rule_add(&ctx, NFT_MSG_NEWRULE,
+   trans = nft_trans_rule_add(&ctx, NFT_MSG_DELRULE,
          old_rule);
    if (trans == NULL) {
     err = -ENOMEM;
-- 
1.7.1